src/chat/Eliza.cpp

```diff
@@ -0,0 +1,421 @@
+    if (overflow <= 0) {
+        return "potential circular cont (@) jump in databases";
+    }
+    return response_buf;
+}
+
+bool Eliza::load_databases(const char *file, char recurflag) {
+    if (num_databases_ >= MaxDatabases)
+        return false;
+
+    FILE *data;
+    char str[MaxInputLength];
+    if ((data = fopen(file, "rt")) == nullptr) {
+        return false;
+    }
+    int linecount = 0;
+
+    while (fgets(str, MaxInputLength - 1, data) != nullptr) {
+        linecount++;
+        trim(str);
+        if (strlen(str) > 0) {
+            if (str[0] >= '1' && str[0] <= '9') {
+                databases_[num_databases_].current_record()->keyword_responses.add_response(str[0] - '0', &(str[1]));
+            } else
+                switch (str[0]) {
+                case '\0': break;
+                case '(':
+                    databases_[num_databases_].add_record();
+                    databases_[num_databases_].current_record()->keyword_responses.add_keywords(str);
+                    break;
+                case '#': break;
+                case '"': fprintf(stderr, "%s\n", &str[1]); break;
+                case '\'': fprintf(stdout, "%s\n", &str[1]); break;
+                case '>': // add another database
+                    if (num_databases_ < MaxDatabases - 1)
+                        num_databases_++;
+                    else
+                        fputs("response database is full. numdbases>maxdbases", stderr);
+                    if (!register_database_names(&str[1], num_databases_)) {
+                        puts("name database is full.");
+                    }
+                    break;
+                case '%': // include another file inline
+                    trim(str + 1);
+                    if (!load_databases(str + 1, 1)) // recurflag set
+                        printf("including inline '%s' failed at %d!\n", str + 1, linecount);
+                    break;
+                case '@': // link the current database to an earlier database to reuse its entries.
+                    if (databases_[num_databases_].linked_database_num == NoDatabaseLink) {
+                        trim(str + 1);
+                        for (char *i = str + 1; *i != 0; i++)
+                            if (*i == '_')
+                                *i = ' ';
+                        sort_databases_by_name();
+                        int linked_db_num = get_database_num_by_exact_name(str + 1);
+                        databases_[num_databases_].linked_database_num = linked_db_num;
+                    } else
+                        printf(" @ database already linked %d at %d\n", num_databases_, linecount);
+                    break;
+                default: printf("extraneous line: '%s' at %d\n", str, linecount);
+                }
+        }
+    }
+    fclose(data);
+    if (!recurflag) {
+        num_databases_++;
+        sort_databases_by_name();
+    }
+    return true;
+}
+
+char *Eliza::trim(char str[]) {
```

Comment:
Would be ace to rephrase these using the string_utils. I think this does a two-sided trim? We don't have that yet, in fairness. This is also in-place.

Code like this makes my eyes cross :grinning: